### PR TITLE
feat: piper-g2p 独立 G2P パッケージ (Python + Rust + JS/WASM)

### PR DESCRIPTION
**Pull request by ayutaz**
## Summary

eSpeak-ng (GPL-3.0) に依存しない、MIT ライセンスの多言語 G2P パッケージエコシステムを新規作成。3 プラットフォーム × 8 言語対応。

- **`piper-g2p` (Python/PyPI)** — 8 言語、IPA-first 設計、PiperEncoder (strict mode)、375 テスト
- **`piper-g2p` (Rust/crates.io)** — 8 言語、言語別 feature flags、C FFI 層、390 テスト
- **`@piper-plus/g2p` (JS/npm)** — 8 言語、OpenJTalk WASM DI、tree-shaking 対応、380 テスト

**284 files changed, 80,082 insertions(+)**

## 設計方針

- **IPA-first**: `phonemize()` は純粋な IPA トークン列を返す（BOS/EOS/PUA なし）
- **エンコード分離**: PUA マッピング・phoneme_id_map・BOS/EOS は `Encoder` が担当
- **GPL-free**: 全依存が MIT/Apache-2.0/BSD-3-Clause
- **PUA compat version**: 3 パッケージ間の互換性を `pua_compat_version = 1` で管理
- **クロスプラットフォーム整合性**: 共有テストフィクスチャで Python/Rust/JS の出力を検証

### 3プラットフォーム共通
- 8 言語の G2P (JA/EN/ZH/KO/ES/FR/PT/SV)
- `PiperEncoder` / `Encoder` (BOS/EOS/PAD 挿入 + PUA マッピング)
- Strict mode (未知トークンでエラー or スキップを選択)
- PUA 互換性バージョンチェック
- カスタム辞書対応
- 多言語テキストの言語自動検出 + セグメンテーション

### Python 固有
- `segment_text()` API (JS 版と同等)
- デフォルト multilingual phoneme_id_map 同梱
- ベンチマークスイート
- mypy + ruff (lint/format) によるコード品質保証

### Rust 固有
- C FFI 層 (`extern "C"`) — iOS/Android 向け UniFFI バインディングの基盤
- `TextSegment` 構造体による構造化セグメンテーション
- clippy + cargo fmt による品質保証

### JS/WASM 固有
- OpenJTalk WASM によるブラウザ内日本語 G2P
- Node.js test runner による 380 テスト

### CI/CD (全ワークフロー緑)
- `g2p-python-ci.yml` — ruff lint + mypy + pytest (3 OS × 3 Python)
- `g2p-rust-ci.yml` — clippy + fmt + test + build (3 OS)
- `g2p-wasm-ci.yml` — Node.js テスト (3 OS) + size check
- `g2p-cross-platform-ci.yml` — Python/Rust/JS 整合性検証

### ドキュメント
- 3 パッケージの THIRD_PARTY_LICENSES.md
- README に Piper 互換性・クロスプラットフォーム情報追加
- ゴールデンテスト生成スクリプト

## Test plan

- [x] Python: `cd src/python/g2p && uv run pytest tests/ -q` (375 tests)
- [x] Rust: `cd src/rust && cargo test -p piper-g2p --all-features` (390 tests)
- [x] JS/WASM: `cd src/wasm/g2p && node --test 'test/test-*.js'` (380 tests)
- [x] Cross-platform consistency: 共有フィクスチャによる出力一致検証
- [x] piper-core 互換: `cd src/rust && cargo test -p piper-plus`
- [x] CI 全 16 ワークフロー緑

### CODE REVIEW

File: src/python/g2p/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,81 @@
+[build-system]
+requires = ["hatchling"]
+build-backend = "hatchling.build"
+
+[project]
+name = "piper-g2p"
+version = "0.1.0"
+description = "Multilingual G2P (Grapheme-to-Phoneme) for TTS — eSpeak-ng free, MIT licensed"
+requires-python = ">=3.11"
+license = "MIT"
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/g2p/pyproject.toml]: Fixed in c9f685c. Changed to `license = { text = "MIT" }` for PEP 621 compliance.

File: src/python/g2p/piper_plus_g2p/encode/encoder.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+    def _tokens_to_raw_ids(self, tokens: list[str]) -> list[int]:
+        """Map token strings -> flat list of phoneme IDs (no padding)."""
+        ids: list[int] = []
+        for token in tokens:
+            mapped = map_token(token)
+            for ch in mapped:
+                if ch in self._id_map:
+                    ids.extend(self._id_map[ch])
+        return ids
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/g2p/piper_plus_g2p/encode/encoder.py:125]: Fixed in c9f685c. Added `_log.warning()` for unknown tokens instead of silent drop.

File: src/python/g2p/piper_plus_g2p/japanese.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        return "ja"
+
+    def phonemize(self, text: str) -> list[str]:
+        tokens, _prosody = _phonemize_core(text)
+        return tokens
+
+    def phonemize_with_prosody(
+        self, text: str
+    ) -> tuple[list[str], list[ProsodyInfo | None]]:
+        return _phonemize_core(text)
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/g2p/piper_plus_g2p/japanese.py:246]: Fixed in c9f685c. Added `self._sanitize_input(text)` at the top of both methods with early return for empty input.

File: src/python/g2p/piper_g2p/encode/pua.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+__all__ = ["FIXED_PUA_MAPPING", "TOKEN2CHAR", "CHAR2TOKEN", "map_token"]
+
+_log = logging.getLogger(__name__)
+
+_PUA_JSON = Path(__file__).parent.parent / "data" / "pua.json"
+
+
+def _load_pua_mapping() -> dict[str, int]:
+    """Load PUA mapping from the canonical JSON file."""
+    with open(_PUA_JSON) as f:
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/g2p/piper_g2p/encode/pua.py]: Fixed in c9f685c. Switched to `importlib.resources.files()` and added `encoding=utf-8`.

File: src/python/g2p/piper_plus_g2p/registry.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+    """
+    # Built-in phonemizers (table-driven)
+    for code, module, class_name in _LANGUAGE_TABLE:
+        try:
+            mod = importlib.import_module(module, package=__package__)
+            cls = getattr(mod, class_name)
+            register_language(code, cls())
+        except ImportError:
+            reason = _SKIP_REASONS.get(code, "missing dependency")
+            _LOGGER.info("Skipping %s: %s", code.upper(), reason)
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/g2p/piper_plus_g2p/registry.py:252]: Fixed in c9f685c. Narrowed to `except ModuleNotFoundError` for expected skips. Unexpected `ImportError` logged at WARNING with `exc_info=True`.

File: src/python/g2p/tests/test_spanish.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
+"""Tests for piper_g2p.spanish -- SpanishPhonemizer."""
+
+import pytest
+
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/g2p/tests/test_spanish.py]: Fixed in c9f685c. Removed unused import.

File: src/python/g2p/tests/test_portuguese.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
+"""Tests for piper_g2p.portuguese -- PortuguesePhonemizer."""
+
+import pytest
+
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/g2p/tests/test_portuguese.py]: Fixed in c9f685c. Removed unused import.

File: src/python/g2p/tests/test_french.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
+"""Tests for piper_g2p.french -- FrenchPhonemizer."""
+
+import pytest
+
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/g2p/tests/test_french.py]: Fixed in c9f685c. Removed unused import.

File: src/python/g2p/tests/test_english.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
+"""Tests for piper_g2p.english — EnglishPhonemizer."""
+
+import pytest
+
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/g2p/tests/test_english.py]: Fixed in c9f685c. Removed unused import.

File: src/python/g2p/tests/test_korean.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
+"""Tests for piper_g2p.korean -- KoreanPhonemizer."""
+
+import pytest
+
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/g2p/tests/test_korean.py]: Fixed in c9f685c. Removed unused import.

File: src/python/g2p/tests/test_chinese.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
+"""Tests for piper_g2p.chinese -- ChinesePhonemizer."""
+
+import pytest
+
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/g2p/tests/test_chinese.py]: Fixed in c9f685c. Removed unused import.

File: src/python/g2p/tests/test_e2e.py [hunk 12]
```diff
@@ -0,0 +1,89 @@
+"""End-to-end tests: text -> phonemize -> encode -> phoneme_ids."""
+
+import pytest
+
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/g2p/tests/test_e2e.py]: Fixed in c9f685c. Removed unused import.

File: src/python/g2p/tests/test_edge_cases.py [hunk 13]
```diff
@@ -0,0 +1,275 @@
+"""Edge case tests for piper_g2p phonemizers."""
+
+import pytest
+
+from piper_g2p import get_phonemizer, available_languages
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/g2p/tests/test_edge_cases.py]: Fixed in c9f685c. Removed unused `available_languages` import.

File: src/python/g2p/tests/test_registry.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
+"""Tests for piper_g2p.registry — language phonemizer registry."""
+
+import pytest
+
+from piper_g2p.base import Phonemizer, ProsodyInfo
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/g2p/tests/test_registry.py]: Fixed in c9f685c. Removed unused `ProsodyInfo` import.

File: src/python/g2p/README.md [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+```python
+from piper_g2p import get_phonemizer
+
+ja = get_phonemizer("ja")
+ja.phonemize("こんにちは")
+# -> ["k", "o", "N_n", "n", "i", "ch", "i", "h", "a"]
+
+en = get_phonemizer("en")
+en.phonemize("Hello world")
+# -> ["h", "ʌ", "ˈ", "l", "oʊ", " ", "ˈ", "w", "ɜː", "l", "d"]
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/g2p/README.md]: Fixed in c9f685c. Aligned README examples with snapshot tests (JA uses `w` + prosody marker, EN matches actual output).
